New test data models documentation (#687)
commercetools/test-data: docs/guidelines/creating-new-model.md

# Building a new test data model

> [!IMPORTANT]
> This document builds on top of the [Test Data Models overview](../contributing/test-data-models-overview.md) one so it would be useful if you read it first.

> In this document we explain the steps we need to take in order to build a new test data model.

Reviewer: ```suggestion
In this document we explain the steps needed to build a new test data model.
```

Author: Updated here: 14c1eb20ced9a64b84849173af9dd5ca3058e1f0

## Context

> First of all, remember that a test data model is a set of objects that allows to create objects that mimic commercetools APIs responses which help during testing implementation.

Reviewer 2: ```suggestion
First of all, remember that a test data model is a set of objects that allow users to create objects that mimic commercetools APIs responses which help during testing implementation.
```

Author: Update here: 1f27b84eeec63545fbb7e79301ffcfbdb5d30b36

> & {
  name?: TLocalizedStringDraftGraphql;
  description?: TLocalizedStringDraftGraphql;
  geoLocation?: TGeometryGraphql;
  __typename: 'ChannelDraft';
};

export type TCreateChannelBuilder<
  TChannelModel extends
    | TChannelRest
    | TChannelGraphql
    | TChannelDraftRest
    | TChannelDraftGraphql,
> = () => TBuilder<TChannelModel>;
```

As you can see, we have four main types: two for the full entity (`TChannelRest` and `TChannelGraphql`) and another pair for the draft version (`TChannelDraftRest` and `TChannelDraftGraphql`).

The REST representations are just using the `@commercetools/platform-sdk`.
This is a package which is auto-generated with the latest versions of the commercetools REST APIs and already contains all the Typescript interfaces describing the models.

On the other hand, the GraphQL representation can use the previous ones as a blueprint, but most likely will use different types for some of their properties (specially the nested models).

## Configuring fields

The next step can involve creating the `fields-config` file where we need to implement the population of the default values we want the test data model to have.

> Since we have two representations for it (REST and GRAPHQL), we need to export two objects.

Reviewer: ```suggestion
Since we have two representations for it (REST and GraphQL), we need to export two objects.
```

Author: Updated here: 14c1eb20ced9a64b84849173af9dd5ca3058e1f0

const commonFieldsConfig = {
  id: fake((f) => f.string.uuid()),
  version: sequence(),
  createdAt: fake(getOlderDate),
  createdBy: fake(() => ClientLogging.random()),
  lastModifiedAt: fake(getNewerDate),
  lastModifiedBy: fake(() => ClientLogging.random()),
  key: fake((f) => f.lorem.slug(2)),
  roles: [roles.Primary],
  name: null,
  description: null,
  address: fake(() => Address.random()),
  reviewRatingStatistics: null,
  custom: null,
  geoLocation: null,
};
export const restFieldsConfig: TModelFieldsConfig<TChannelRest> = {
  fields: {
    ...commonFieldsConfig,
  },
};
export const graphqlFieldsConfig: TModelFieldsConfig<TChannelGraphql> = {
  fields: {
    ...commonFieldsConfig,
    __typename: 'Channel',
    nameAllLocales: null,
    descriptionAllLocales: null,
  },
  postBuild: (model) => {
    const name = model.nameAllLocales
      ? LocalizedString.resolveGraphqlDefaultLocaleValue(model.nameAllLocales)
      : undefined;
    const description = model.descriptionAllLocales
      ? LocalizedString.resolveGraphqlDefaultLocaleValue(
          model.descriptionAllLocales
        )
      : undefined;
    return {
      name,
      description,
    };
  },
};
```

> Probably you have spotted the `postBuild` property in the Graphql config object.

Reviewer: ```suggestion
Probably you have spotted the `postBuild` property in the GraphQL config object.
```

Author: Updated here: 14c1eb20ced9a64b84849173af9dd5ca3058e1f0

> This is an optional one which we should only use if there are any dependencies among the data model fields.

Reviewer: ```suggestion
This is an optional and we should only use if there are any dependencies among the data model fields.
```

Author: Updated here: 14c1eb20ced9a64b84849173af9dd5ca3058e1f0

> This is an optional and we should only use if there are any dependencies among the data model fields.
> In this case, `name` and `description` depend on the values from `nameAllLocales` and `descriptionAllLocales` but we don't know them in advance but only once the data model has been built. The `postBuild` callback allows to manipulate the built object before it's returned to the consumer.

Reviewer 3: Nit: Small grammatical cleanup.
```suggestion
This is optional and we should only use it if there are any dependencies among the data model fields.
In this case, `name` and `description` depend on the values from `nameAllLocales` and `descriptionAllLocales`. We won't know the values in advance, but will once the data model has been built. The `postBuild` callback allows us to manipulate the built object before it's returned to the consumer.
```

Author: Thanks for the suggestion, Sarah 🙇 

Applied here: 783806a087a7f9ab2bc517073b570d0d4dab9866
